features/storage/kvstore/tdbstore/DirectAccessDevicekey.cpp

```diff
@@ -0,0 +1,284 @@
+static inline uint32_t align_down(uint64_t val, uint64_t size);
+static uint32_t calc_crc(uint32_t init_crc, uint32_t data_size, const void *data_buf);
+
+// -------------------------------------------------- API Functions Implementation ----------------------------------------------------
+int direct_access_to_devicekey(uint32_t tdb_start_offset, uint32_t tdb_end_offset, void *data_buf,
+                               size_t data_buf_size, size_t *actual_data_size_ptr)
+{
+    int status = MBED_ERROR_INVALID_ARGUMENT;
+    FlashIAP flash;
+    uint8_t active_area = 0;
+    tdbstore_area_data_t area_params[TDBSTORE_NUMBER_OF_AREAS];
+    memset(area_params, 0, sizeof(area_params));
+    bool is_flash_init = false;
+
+    if (NULL == data_buf) {
+        tr_error("Invalid Data Buf Argument");
+        goto exit_point;
+    }
+
+    status = flash.init();
+    if (status != 0) {
+        tr_error("FlashIAP init failed - err: %d", status);
+        status = MBED_ERROR_FAILED_OPERATION;
+        goto exit_point;
+    }
+
+    is_flash_init = true;
+
+    status = calc_area_params(&flash, tdb_start_offset, tdb_end_offset, area_params);
+    if (status != MBED_SUCCESS) {
+        tr_error("Couldn't calulate Area Params - err: %d", status);
+        goto exit_point;
+    }
+
+    /* DeviceKey data can be found either in first or second Flash Area */
+    /* Loop through Areas to find valid DeviceKey data */
+    for (active_area = 0; active_area < TDBSTORE_NUMBER_OF_AREAS; active_area++) {
+        status = reserved_data_get(&flash, &area_params[active_area], data_buf, data_buf_size, actual_data_size_ptr);
+        if (status == MBED_SUCCESS) {
+            break;
+        }
+    }
+
+    if (status != MBED_SUCCESS) {
+        status = MBED_ERROR_ITEM_NOT_FOUND;
+        tr_error("Couldn't find valid DeviceKey - err: %d", status);
+    }
+
+exit_point:
+    if (true == is_flash_init) {
+        flash.deinit();
+        is_flash_init = false;
+    }
+
+    return status;
+}
+
+int  get_expected_internal_TDBStore_position(uint32_t *out_tdb_start_offset, uint32_t *out_tdb_end_offset)
+{
+    uint32_t flash_end_address;
+    uint32_t flash_start_address;
+    uint32_t aligned_start_address;
+    uint32_t aligned_end_address;
+    FlashIAP flash;
+    bool is_default_configuration = false;
+    uint32_t tdb_size;
+
+    int ret = flash.init();
+    if (ret != 0) {
+        return -1;
+    }
+
+    if (strcmp(STR(MBED_CONF_STORAGE_STORAGE_TYPE), "FILESYSTEM") == 0) {
+        *out_tdb_start_offset =  MBED_CONF_STORAGE_FILESYSTEM_INTERNAL_BASE_ADDRESS;
+        tdb_size = MBED_CONF_STORAGE_FILESYSTEM_RBP_INTERNAL_SIZE;
+    } else if (strcmp(STR(MBED_CONF_STORAGE_STORAGE_TYPE), "TDB_EXTERNAL") == 0) {
+        *out_tdb_start_offset =  MBED_CONF_STORAGE_TDB_EXTERNAL_INTERNAL_BASE_ADDRESS;
+        tdb_size = MBED_CONF_STORAGE_TDB_EXTERNAL_RBP_INTERNAL_SIZE;
+    } else {
+        return MBED_ERROR_UNSUPPORTED;
```

Comment:
Should deinit flash here and on next failures.

Reply:
my bad, some of the ifs are later addition to the code. I will move flash init below them.